tls/s2n_handshake_io.c

```diff
@@ -1040,27 +1015,41 @@ int s2n_negotiate(struct s2n_connection *conn, s2n_blocked_status * blocked)
         this = 'C';
     }
 
-    while (ACTIVE_STATE(conn).writer != 'B') {
-        /* Flush any pending I/O or alert messages */
-        GUARD(s2n_flush(conn, blocked));
+    /* If we were blocked reading a record, call the handler right away. Don't try to read more data. */
+    if (conn->handshake.blocked_on_read) {
+        int r = ACTIVE_STATE(conn).handler[conn->mode] (conn);
 
-        if (ACTIVE_STATE(conn).writer == 'A') {
-            /* We are in a state that is blocked on application data */
-            *blocked = S2N_BLOCKED_ON_APPLICATION_INPUT;
+        if (r < 0) {
+            /* If we are still blocked, return control to the caller. */
+            if (s2n_errno == S2N_ERR_BLOCKED) {
+                S2N_ERROR_PRESERVE_ERRNO();
+            }
 
-            int r = s2n_handshake_handle_app_data(conn);
-            if (r < 0) {
+            /* Otherwise there is some other problem and we should kill the connection. */
+            if (s2n_allowed_to_cache_connection(conn) && conn->session_id_len) {
                 s2n_try_delete_session_cache(conn);
-                /* The peer might have sent an alert. Try and read it. */
-                if (s2n_errno == S2N_ERR_BLOCKED && s2n_stuffer_data_available(&conn->in)) {
-                    if (handshake_read_io(conn) < 0 && s2n_errno == S2N_ERR_ALERT) {
-                        /* handshake_read_io has set s2n_errno */
-                        S2N_ERROR_PRESERVE_ERRNO();
-                    }
-                }
-                S2N_ERROR_PRESERVE_ERRNO();
             }
-        } else if (ACTIVE_STATE(conn).writer == this) {
+
+            GUARD(s2n_connection_kill(conn));
+            S2N_ERROR_PRESERVE_ERRNO();
+        }
+
+        /* The handler completed successfully, we are done with this record. */
+        /* Advance the state machine and wipe the record. */
+        GUARD(s2n_advance_message(conn));
```

Comment:
This looks wired as `handshake_read_io` would advance the state machine. It looks like `s2n_establish_session` works independently and is not processed by `handshake_read_io`.

Reply:
This works very similar to the `handle_app_handshake_data` function which also advanced the state machine. Since we don't have 'A' states in this scenario, `handshake_read_io` fails with a blocking error, and never advances the state machine. If we detect that the previous message was `blocked_on_read`, then we jump back into the message handler. I prefer the AsyncTLS terminology here. `blocked_on_read` should be `paused` or something similar. That makes it more clear that every time we have a paused handshake, we just pick up where we left off.